tools/autograd/derivatives.yaml

```diff
@@ -1010,7 +1010,13 @@
   save_invstd: not_implemented("native_batch_norm_backward save_invstd")
 
 - name: native_layer_norm(Tensor input, int[] normalized_shape, Tensor? weight, Tensor? bias, float eps) -> (Tensor, Tensor, Tensor)
-  input, weight, bias: "GradMode::is_enabled() || grads[1].defined() || grads[2].defined() ? infinitely_differentiable_native_layer_norm_backward(grads[0], grads[1], grads[2], input, result1, result2, weight, normalized_shape, eps, grad_input_mask) : (grads[0].defined() ? native_layer_norm_backward(grads[0].is_contiguous() ? grads[0] : grads[0].contiguous(), input, normalized_shape, result1, result2, weight, bias, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>())"
+  input, weight, bias: "grad.defined() ? native_layer_norm_backward(grad, input, normalized_shape, result1, result2, weight, bias, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
+
+- name: native_layer_norm_backward(Tensor grad_out, Tensor input, int[] normalized_shape, Tensor mean, Tensor rstd, Tensor? weight, Tensor? bias, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
+  input, weight, grad_out: layer_norm_double_backward(input, weight, grads[0], grads[1], grads[2], grad_out, mean, rstd, normalized_shape, grad_input_mask)
+  bias: Tensor()
```

Comment:
The bias value is not used when evaluating native_layer_norm_backward ? We can remove it then to save memory

Reply:
This one is a little messy.
bias is only used to get dbias options dtype: https://github.com/pytorch/pytorch/blob/289bcac618448b04bd68a2acd577c0261e368da6/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L555

Technically weight & bias should have the same dtype. But weight/bias are independent and weight could be undefined.
I can remove bias and replace it with an optional dtype.

Reply:
Yes replacing it with dtype is better as the codegen will stop saving the full bias Tensor for no reason.

Reply:
Do we want to do that in this PR as well?
This one has been dragging for a while and I would really like to have the two autodiff PRs merged sooner than later.

I can start another PR for the API change at the same time.

Reply:
Turns out it's quite sticky to update that in derivatives.yaml: https://github.com/pytorch/pytorch/issues/62934